17-lifecycle-lab/src/01-cps.jsx

```diff
@@ -7,12 +7,63 @@
  */
 
 import React from 'react';
+import CPSIndicator from './02-cps-with-indicator.jsx'
 
-export default createClass({
-  render() {
-    return (<div>
-      <button>Click Fast</button>
-      <p>CPS rate: {0}</p>
-    </div>);
-  }
-});
+export default class CPS extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            numberOfClicks: 0,
+            seconds: 0
+        }
+        this.calcNumOfClicks.bind(this);
+    }
+
+    componentWillUnmount() {
+        clearInterval(this.timer);
+    }
+
+    componentWillMount() {
+        this.start =
+            this.timer = setInterval(this.tick, 1000);
+    }
+
+    tick = () => {
+        let seconds = this.state.seconds + 1;
+        this.setState({seconds});
+    }
+
+    clicked = () => {
+        let numberOfClicks = this.state.numberOfClicks + 1;
+        this.setState({numberOfClicks});
+    }
+
+
+    calcNumOfClicks() {
+        return (this.state.numberOfClicks / this.timer);
+    }
+
+    render() {
+        const numOfClicks = this.state.numberOfClicks;
+        const sec = this.state.seconds;
+        const frequency = Number(numOfClicks / sec);
+        return (<div>
+            <button onClick={this.clicked}>Click Fast</button>
+            <p>CPS rate: {frequency}</p>
+
+            {frequency < 1 &&
+                <div>
+                    <p>Too Slow</p>
+                    <CPSIndicator color="red" />
```

Comment:
better instead of passing color, pass valid or invalid prop. let the indicator itself define the proper color. Parent component should not be aware of indicator color.